### PR TITLE
Status development

### PR DESCRIPTION
**Pull request by Antonhansel**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+.DS_Store
+nodes_modules
```
Comment (Amoki) [inline .gitignore]: nodes_modules/* ?

File: server.js [hunk 2]
```diff
@@ -1,13 +1,19 @@
...
 // get all the tools we need
-var express  	 = require('express');
-var app      	 = express();
-var port     	 = process.env.PORT || 8080;
-var morgan       = require('morgan');
-var path 		 = require('path');
+var morgan       	   = require('morgan');
+var path 		  	     = require('path');
+var webSocketServer  = require('ws').Server;
+var http 		     	   = require('http');
```
Comment (Amoki) [inline server.js]: spaces and tabs

File: server.js [hunk 3]
```diff
@@ -1,13 +1,19 @@
...
+var express 		     = require('express');
+var app              = express();
+var server           = http.createServer(app);
+var wss              = new webSocketServer({server: server});
+console.log('WebSocket server online, listening on ' + config.port);
 // configuration ===============================================================
 // set up our express application
+server.listen(config.port);
```
Comment (Amoki) [inline server.js]: server.listen(config.port, function () {
  console.log('WebSocket server online, listening on ' + config.port);
});

File: README.md [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# BitData.io
+# Status Fetch
```
Comment (Neamar) [inline README.md]: status.anyfetch.com

File: README.md [hunk 5]
```diff
@@ -1,4 +1,4 @@
...
```
Comment (Neamar) [inline README.md:18]: Say a little more about the project

File: bower.json [hunk 6]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "status.anyfetch.com",
+  "version": "0.0.0",
+  "homepage": "https://github.com/AnyFetch/status.anyfetch.com",
+  "authors": [
+    "antonhansel <antoninribeaud@gmail.com>"
+  ],
+  "main": "server.js",
+  "moduleType": [
```
Comment (Neamar) [inline bower.json]: Add a script key for linting

File: lib/app/crawler.js [hunk 7]
```diff
@@ -1,10 +1,30 @@
...
-	request.get('url', function(err, res){
-		if (err) console.log(err);
-		else marketCap = JSON.parse(res.body);
-	});
+var request = require('supertest');
+var async = require('async');
+
+var savedData = [];
+
+exports.pollData = function(socket, urls, timeout) {
```
Comment (Neamar) [inline lib/app/crawler.js]: use module.exports, avoid tricky bugs ;)

Comment (Neamar) [inline lib/app/crawler.js]: name the function

File: lib/app/crawler.js [hunk 8]
```diff
@@ -1,10 +1,30 @@
...
+  var results = {};
+
+  async.map(urls, function(url, cb) {
+    request(url[1])
```
Comment (Neamar) [inline lib/app/crawler.js]: why url[1] ? Can you use a key?

File: lib/app/crawler.js [hunk 9]
```diff
@@ -1,10 +1,30 @@
...
+    .get('/status')
+    .timeout(timeout)
+    .expect(200)
+    .end(function(err, res) {
+      results[url[0]] =  err ? -1 : res.body;
```
Comment (Neamar) [inline lib/app/crawler.js]: idem, use a key instead of url[0]

File: lib/config/config.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
+module.exports = {
+  "env": "production",
+  "timeout": 1000,
+  "pollingInterval": 2000,
+  "port": process.env.PORT || 8080,
+  "productionUrls": [
+    ['gmail' , 'https://gmail.anyfetch.com/'],
```
Comment (Neamar) [inline lib/config/config.js]: why don't you use a {} ? `{gmail: 'http://gmail.anyfetch.com'}`

File: lib/app/socketHandler.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
+var io = require('socket.io').listen(8081);
+
+var crawler = require('./crawler.js');
+var intervalProduction;
+var userCount = 0;
+
+
+module.exports = function(config) {
```
Comment (Neamar) [inline lib/app/socketHandler.js]: name the function

File: lib/app/socketHandler.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+  io.on('connection', function(socket) {
+    socket.emit('urls', config.urls);
+    userCount++;
+
+    if (userCount !==1) {
```
Comment (Neamar) [inline lib/app/socketHandler.js]: linter ;)

File: lib/config/configuration.js [hunk 13]
```diff
@@ -0,0 +1,10 @@
+var config = require('./config.js');
+
+if(config.env === "production") {
```
Comment (Neamar) [inline lib/config/configuration.js]: config.urls = config.urls[config.env]

**Comment by Neamar** [comment]
Let's clean this mess -- https://help.github.com/articles/remove-sensitive-data/

Previous: https://github.com/AnyFetch/status.anyfetch.com/commit/90f44fbc809cafa5577d07b8b85446fee97863b5

File: .gitignore [hunk 14]
```diff
@@ -0,0 +1,4 @@
+.DS_Store
+node_modules/*
+**/components
```
Comment (Neamar) [inline .gitignore]: public/components

File: README.md [hunk 15]
```diff
@@ -1,5 +1,20 @@
-# BitData.io
+# status.anyfetch.com
 
+You can use this tool to display of the current load on anyfetch providers, in realtime.
```
Comment (Neamar) [inline README.md]: display the

File: README.md [hunk 16]
```diff
@@ -1,5 +1,20 @@
...
+
+## How to use
+```
+npm install
+node bin/server
+```
+Point your browser on http://localhost:8080
+
+## Configuration
+To add an url or to switch environment, edit config/config.js
```
Comment (Neamar) [inline README.md]: edit `config/config.js` (backtick)

File: README.md [hunk 17]
```diff
@@ -1,5 +1,20 @@
...
+
+The server will generate the statistics and change the page name for you
+
+To add a new environment behavior, edit config/index.js
 License
```
Comment (Neamar) [inline README.md:19]: \n

File: app.js [hunk 18]
```diff
@@ -0,0 +1,22 @@
+// app.js
+// set up ======================================================================
+// get all the tools we need
+var config = require('./config');
+var morgan = require('morgan');
+var path = require('path');
+var express = require('express');
+var app = express.createServer();
+app.listen(config.port);
+var io = require('socket.io').listen(app);
```
Comment (Neamar) [inline app.js:11]: do all the require on top, then \n

File: app.js [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+// configuration ===============================================================
+app.use(morgan('dev')); // log every request to the console
```
Comment (Neamar) [inline app.js]: try to add comments on the line before when you can

File: app.js [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+app.set('view engine', 'ejs'); // set up ejs for templating
+
+// required for passport
+app.use(express.static(path.join(__dirname, 'public'))); // to get local files
+
+// routes ======================================================================
+require('./lib/app/routes.js')(app, config.env); // load our routes and pass in our app
```
Comment (Neamar) [inline app.js]: idem, require on top

File: app.js [hunk 21]
```diff
@@ -0,0 +1,22 @@
+// app.js
+// set up ======================================================================
+// get all the tools we need
+var config = require('./config');
+var morgan = require('morgan');
+var path = require('path');
+var express = require('express');
+var app = express.createServer();
+app.listen(config.port);
```
Comment (Neamar) [inline app.js:10]: the listening should be done on bin/Server -- that way app.js can be required for test and scripts

Comment (Neamar) [inline app.js:10]: See https://github.com/AnyFetch/api.anyfetch.com/blob/master/bin/server

File: bin/server [hunk 22]
```diff
@@ -0,0 +1,3 @@
+// Load appjs
+app = require('../app.js');
+
```
Comment (Neamar) [inline bin/server]: Listening here

File: bin/server [hunk 23]
```diff
@@ -0,0 +1,3 @@
+// Load appjs
+app = require('../app.js');
```
Comment (Neamar) [inline bin/server]: var app

File: bin/server [hunk 24]
```diff
@@ -0,0 +1,3 @@
+// Load appjs
```
Comment (Neamar) [inline bin/server]: "use strict";

File: config/config.js [hunk 25]
```diff
@@ -0,0 +1,28 @@
+module.exports = {
+  "env": process.env.NODE_ENV || "staging",
```
Comment (Neamar) [inline config/config.js]: OQ you don't need to add " " around your keys

File: config/index.js [hunk 26]
```diff
@@ -0,0 +1,5 @@
+var config = require('./config.js');
```
Comment (Neamar) [inline config/index.js]: "use strict";

File: lib/app/crawler.js [hunk 27]
```diff
@@ -1,10 +1,32 @@
...
+    .get('/status')
+    .timeout(timeout)
+    .expect(200)
+    .end(function(err, res) {
+      results[url] =  err ? -1 : res.body;
+      cb();
+    });
+  }, function() {
+    savedData.push(results);
+    if(savedData.length > 100) {
```
Comment (Neamar) [inline lib/app/crawler.js:24]: explain why?

File: lib/app/routes.js [hunk 28]
```diff
@@ -1,13 +1,9 @@
 // app/routes.js
```
Comment (Neamar) [inline lib/app/routes.js]: "use strict";

**Comment by Neamar** [comment]
For filename, we like to follow the Node "-" convention : `lib/app/socketHandler.js` for instance should be `lib/app/socket-handler.js`.

File: lib/app/socketHandler.js [hunk 29]
```diff
@@ -0,0 +1,31 @@
+'use strict';
+
+var crawler = require('./crawler.js');
+var intervalProduction;
+var userCount = 0;
+
+
+module.exports = function(config, io) {
```
Comment (Neamar) [inline lib/app/socketHandler.js]: function name

File: package.json [hunk 30]
```diff
@@ -1,27 +1,36 @@
...
   "version": "0.0.0",
-  "description": "statusfetch",
+  "description": "status.anyfetch.com",
   "main": "server.js",
   "scripts": {
-    "test": "mocha test"
+    "start": "node bin/server",
+    "test": "mocha test",
+    "postinstall": "./node_modules/bower/bin/bower install; npm run-script make",
+    "lint": "jshint lib/app app.js && jscs lib/app/ app.js"
```
Comment (Neamar) [inline package.json]: jshint bin lib/ config/ public/ views app.js 

(same for jscs)

File: package.json [hunk 31]
```diff
@@ -1,27 +1,36 @@
...
   },
   "repository": {
     "type": "git",
-    "url": "https://github.com/Antonhansel/bitdata.git"
+    "url": "https://github.com/anyfetch/status.anyfetch.com.git"
```
Comment (Neamar) [inline package.json]: AnyFetch (i think)

File: package.json [hunk 32]
```diff
@@ -1,27 +1,36 @@
...
   },
-  "author": "",
+  "author": "Antonhansel",
   "license": "ISC",
   "bugs": {
-    "url": "https://github.com/Antonhansel/bitdata/issues"
+    "url": "https://github.com/anyfetch/status.anyfetch.com/issues"
```
Comment (Neamar) [inline package.json]: AnyFetch

File: package.json [hunk 33]
```diff
@@ -1,27 +1,36 @@
...
   },
-  "homepage": "https://github.com/Antonhansel/bitdata",
+  "homepage": "https://github.com/Anyfetch/status.anyfetch.com",
```
Comment (Neamar) [inline package.json:20]: you can leave empty and it will be added automatically on publich (tip)

File: package.json [hunk 34]
```diff
@@ -1,27 +1,36 @@
...
   "dependencies": {
     "async": "^0.9.0",
-    "express": "^4.11.2",
+    "bower": "^1.3.12",
+    "ejs": "^2.2.4",
+    "express": "2.5.5",
+    "http": "0.0.0",
```
Comment (Neamar) [inline package.json]: i think it is native?

File: package.json [hunk 35]
```diff
@@ -1,27 +1,36 @@
...
     "morgan": "^1.5.1",
     "path": "^0.11.14",
-    "request": "^2.53.0"
+    "request": "^2.53.0",
```
Comment (Neamar) [inline package.json]: are you still using it?

File: package.json [hunk 36]
```diff
@@ -1,27 +1,36 @@
...
+    "socket.io": "^1.3.3",
+    "supertest": "^0.15.0",
+    "ws": "^0.7.1"
   },
   "devDependencies": {
     "chai": "^1.10.0",
```
Comment (Neamar) [inline package.json]: not used?

**Comment by Neamar** [comment]
Idem for public/js/fetchData.js => public/js/fetch-data.js

File: public/js/fetchData.js [hunk 37]
```diff
@@ -0,0 +1,78 @@
+var urlsList = null;
+var storedValues = 100;
+
+jQuery(function() {
```
Comment (Neamar) [inline public/js/fetchData.js]: jQuery? Why not $?

File: public/js/fetchData.js [hunk 38]
```diff
@@ -0,0 +1,78 @@
...
+  var socket = io.connect();
+
+  socket.on('savedData', function(res) {
+    var date = new Date().getTime();
+    $('#graph-container').append("<img src=loading.gif id='loading-image' style='align:center'>");
```
Comment (Neamar) [inline public/js/fetchData.js]: loading.gif is not part of the repo

File: public/js/fetchData.js [hunk 39]
```diff
@@ -0,0 +1,78 @@
...
+  $('#graph-container').empty();
+  for(var i = 0; i < dataArray.length; i++) {
+    $('#graph-container').append(dataArray[i].html);
+  }
+});
+
+//
+// Slider Handler
+//
+
```
Comment (Neamar) [inline public/js/fetchData.js]: - \n

File: public/js/graph.js [hunk 40]
```diff
@@ -0,0 +1,138 @@
...
+      html: ''
+    };
+    dataSet.html = generateHtml(dataSet);
+    $('#graph-container').append(dataSet.html);
+    dataArray.push(dataSet);
+  });
+}
+
+function updateAll() {
+  for(var i = 0; i < dataArray.length; i++) {
```
Comment (Neamar) [inline public/js/graph.js]: use a forEach for simple iteration -- free closure + improved readability

File: public/js/graph.js [hunk 41]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+function generateHtml(dataSet) {
+  return ("<div class='panel panel-default'><div class='panel-heading'>" +
+    dataSet.name + "</div><div id=flot-" + dataSet.id +
+    " style='height:300px;'></div></div>");
+}
+
+function updateGraphs(data, date, realtime) {
+  for(var i = 0; i < dataArray.length; i++) {
```
Comment (Neamar) [inline public/js/graph.js]: .forEach

File: public/js/graph.js [hunk 42]
```diff
@@ -0,0 +1,138 @@
...
+    if(data[dataArray[i].id].pending_documents > -1) {
+      dataArray[i].dataSet[0].data.push([date, data[dataArray[i].id].pending_documents]);
+    }
+    else {
+      var found = false;
+      if(warnings.length === 0) {
+        warnings.push([dataArray[i].name, 5]);
+      }
+      else {
+        for(var j = 0; j < warnings.length; j++) {
```
Comment (Neamar) [inline public/js/graph.js]: use .some()

File: public/js/graph.js [hunk 43]
```diff
@@ -0,0 +1,138 @@
...
+    while(dataArray[i].dataSet[0].data.length > storedValues) {
+      dataArray[i].dataSet[0].data.shift();
+    }
+    if(data[dataArray[i].id].pending_documents > -1) {
+      dataArray[i].dataSet[0].data.push([date, data[dataArray[i].id].pending_documents]);
+    }
+    else {
+      var found = false;
+      if(warnings.length === 0) {
+        warnings.push([dataArray[i].name, 5]);
```
Comment (Neamar) [inline public/js/graph.js]: ?What is the 5?

File: public/js/graph.js [hunk 44]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+function showWarnings() {
+  $('#critical-status').empty();
+  if(warnings.length !== 0) {
+    $('#critical-status').append('<p>Not Responding: ' + warnings.length + '</p>');
+  }
+  else {
+    $('#critical-status').append('<p>Everything looks good!</p>Documents pending:</br>');
+    for(var i = 0; i < dataArray.length; i++) {
```
Comment (Neamar) [inline public/js/graph.js]: forEach

File: views/index.ejs [hunk 45]
```diff
@@ -1,19 +1,37 @@
...
-    <script src="http://code.highcharts.com/modules/exporting.js"></script>
-    <link rel="stylesheet" href="//netdna.bootstrapcdn.com/font-awesome/4.0.3/css/font-awesome.min.css"> <!-- load fontawesome -->
-    <link rel="stylesheet" href="dist/css/bootstrap.min.css">
-    <style>
-        body {
-            padding-top:80px;
-            background-color: #171d25;
-        }
-    </style>
+    <title>Status Fetch v1.0</title>
```
Comment (Neamar) [inline views/index.ejs]: Status AnyFetch ;)

File: views/index.ejs [hunk 46]
```diff
@@ -1,19 +1,37 @@
...
+    <link rel="stylesheet" href="https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/themes/smoothness/jquery-ui.css"/>
```
Comment (Neamar) [inline views/index.ejs]: Use // instead of https (we may serve this over https later)

File: views/index.ejs [hunk 47]
```diff
@@ -1,19 +1,37 @@
...
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.2/css/bootstrap.min.css"/>
```
Comment (Neamar) [inline views/index.ejs]: Use // instead of https (we may serve this over https later)

File: views/index.ejs [hunk 48]
```diff
@@ -1,19 +1,37 @@
...
+                    <p id="grid-mode"><span class="glyphicon glyphicon-th-large"></span></p>
+                    <p id="justify-mode"><span class="glyphicon glyphicon-align-justify"></span></p>
+                </p>
+            </div>
+        </div>
+        <div class="jumbotron" id='critical-status'></div>
+        <div class="row" id="graph-container"></div>
+    </div>
 </body>
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
```
Comment (Neamar) [inline views/index.ejs]: Use // instead of https (we may serve this over https later)

File: views/index.ejs [hunk 49]
```diff
@@ -1,19 +1,37 @@
...
+<script src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/jquery-ui.min.js"></script>
```
Comment (Neamar) [inline views/index.ejs]: Use // instead of https (we may serve this over https later)

File: views/index.ejs [hunk 50]
```diff
@@ -1,19 +1,37 @@
...
+<script src="https://cdn.socket.io/socket.io-1.3.3.js"></script>
```
Comment (Neamar) [inline views/index.ejs]: Use // instead of https (we may serve this over https later)

File: views/index.ejs [hunk 51]
```diff
@@ -1,19 +1,37 @@
...
+<script src="components/flot/jquery.flot.js"></script>
+<script src="components/flot/jquery.flot.time.js"></script>
+<script src="components/flot-axislabels/jquery.flot.axislabels.js"></script>
+<script src="js/fetchData.js"></script>
+<script src="js/graph.js"></script>
+<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.2/js/bootstrap.min.js"></script>
```
Comment (Neamar) [inline views/index.ejs]: Use // instead of https (we may serve this over https later)

Comment (Neamar) [inline views/index.ejs]: Also, do you use bootstrap js?

File: public/css/index.css [hunk 52]
```diff
@@ -0,0 +1,12 @@
+.legendLabel
+{
+	color:White;
```
Comment (Neamar) [inline public/css/index.css]: White should be white according to CSS rules i believe

File: public/css/index.css [hunk 53]
```diff
@@ -0,0 +1,12 @@
...
+}
+.custom-body {
+	padding-top:80px;
```
Comment (Neamar) [inline public/css/index.css:6]: tabs used instead of spaces

File: views/index.ejs [hunk 54]
```diff
@@ -1,19 +1,37 @@
...
+                <div id="slider"></div>
+            </div>
+            <div class="col-md-4">
+                <p>Change layout
+                    <p id="grid-mode"><span class="glyphicon glyphicon-th-large"></span></p>
+                    <p id="justify-mode"><span class="glyphicon glyphicon-align-justify"></span></p>
+                </p>
+            </div>
+        </div>
+        <div class="jumbotron" id='critical-status'></div>
```
Comment (Neamar) [inline views/index.ejs]: ' => "

File: views/index.ejs [hunk 55]
```diff
@@ -1,19 +1,37 @@
...
-    <style>
-        body {
-            padding-top:80px;
-            background-color: #171d25;
-        }
-    </style>
+    <title>Status AnyFetch v1.0</title>
+    <link rel="stylesheet" href="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/themes/smoothness/jquery-ui.css"/>
+    <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootstrap/3.3.2/css/bootstrap.min.css"/>
+    <link rel='stylesheet' href='css/index.css'>
```
Comment (Neamar) [inline views/index.ejs]: ' => "

File: app.js (hunk 21 above)
Comment (Antonhansel) [inline app.js:10]: I cannot do this given the special configuration of my app.
Listening need to be done earlier.

**Pull request merged**